keps/sig-api-machinery/20200309-consistent-resource-versions-semantics.md

```diff
@@ -0,0 +1,216 @@
+	// older than this ResourceVersion may be served.
+	// This guarantees that ResourceVersion in the ListMeta is not older than the requested
+	// ResourceVersion, but does not make any guarantee about the ResourceVersion in the ObjectMeta
+	// of the list items since ObjectMeta.ResourceVersion tracks when an object was last updated,
+	// not how up-to-date the object is when served.
+	ResourceVersionMatchNotOlderThan ResourceVersionMatch = "NotOlderThan"
+	// ResourceVersionMatchExact matches data at the exact ResourceVersion
+	// provided. If the provided ResourceVersion is unavailable, the server responds with
+	// HTTP 410 “Gone”.
+	// This guarantees that ResourceVersion in the ListMeta is the same as the requested
+	// ResourceVersion, but does not make any guarantee about the ResourceVersion in the ObjectMeta
+	// of the list items since ObjectMeta.ResourceVersion tracks when an object was last updated,
+	// not how up-to-date the object is when served.
+	ResourceVersionMatchExact ResourceVersionMatch = "Exact"
+)
+```
+
+```
+type ListOptions struct {
+    ...
+    // When specified with a watch call, shows changes that occur after that particular version of a resource.
+	// Defaults to changes from the beginning of history.
+	// If set for a list call, ResourceVersionMatch should also be set.
+	// When specified for list:
+	// - if unset, then the result is returned from remote storage based on quorum-read flag;
+	// - if set and ResourceVersionMatch is set, requests that the server apply the ResourceVersionMatch rule;
+	// - if set and ResourceVersionMatch is unset or the server ignores ResourceVersionMatch, the legacy behavior applies:
+	//   - if 0, the result may contain arbitrarily old data, no guarantee;
+	//   - if non-zero and Limit is unset, ResourceVersionMatchNotOlderThan rule applies implicitly;
+	//   - if non-zero and Limit is set, ResourceVersionMatchExact rules applies implicitly.
+	// +optional
+	ResourceVersion string `json:"resourceVersion,omitempty" protobuf:"bytes,4,opt,name=resourceVersion"`
+
+	// ResourceVersionMatch determines how ResourceVersion is applied. Not supported for watch calls.
+	// ResourceVersionMatch SHOULD be set for list calls where ResourceVersion is set. If ResourceVersion is unset,
+	// ResourceVersionMatch is ignored.
+	// For backward compatibility, clients must tolerate the server ignoring ResourceVersionMatch:
+	// - When using ResourceVersionMatchNotOlderThan and Limit is set, clients must handle HTTP 410 “Gone” responses.
+	//   For example, the client might retry with a newer ResourceVersion or fall back to a ResourceVersion="" request.
+	// - When using ResourceVersionMatchExact and Limit is unset, clients must verify that the ResourceVersion in the
+	//   ListMeta of the response matches the requested ResourceVersion, and handle the case where it does not. For
+	//   example, the client might fall back to the a request with Limit set.
+	// +optional
```

Comment:
I think we should mention the performance impact here. The whole reason behind even having this "NotOlderThan" semantic is just for performance/scalability - we can simply serve them much cheaper. So should we mention something about that explicitly?
e.g. Unless you have strong consistency requirements, using NotOlderThan semantic is preferable to achieve better performance and scalability of your cluster?
[or however better it can be worder]

@lavalamp - WDYT?

Reply:
I've added:

```
Unless you have strong consistency requirements, using ResourceVersionMatchNotOlderThan and a known                                                                                                                                                                                                                                                                                                                                  
ResourceVersion is preferable since it can achieve better performance and scalability of your cluster                                                                                                                                                                                                                                                                                                                                than leaving ResourceVersion and ResourceVersionMatch unset, which requires quorum read to be served.
```